backend/de.metas.adempiere.adempiere/base/src/main/java/org/adempiere/archive/api/impl/ArchiveDAO.java

```diff
@@ -141,4 +154,28 @@ public I_AD_Archive retrieveArchive(@NonNull final ArchiveId archiveId)
 	{
 		return InterfaceWrapperHelper.load(archiveId, I_AD_Archive.class);
 	}
+
+	@Override
+	public <T> Stream<AdArchive> streamArchivesForFilter(final IQueryFilter<T> outboundLogFilter, final Class<T> objectClass, final Function<T, TableRecordReference> tableRecordReferenceFunction)
+	{
+		Check.assumeNotNull(outboundLogFilter, "filter cannot not null");
+
+		final IQueryBuilder<T> queryBuilder = queryBL.createQueryBuilder(objectClass)
+				.addOnlyActiveRecordsFilter().filter(outboundLogFilter);
+
+		return queryBuilder.create()
+				.stream()
+				.map(log -> retrieveLastArchives(Env.getCtx(), tableRecordReferenceFunction.apply(log), 1).stream().findFirst())
+				.filter(Optional::isPresent)
+				.map(Optional::get)
+				.map(arch -> AdArchive.builder().id(ArchiveId.ofRepoId(arch.getAD_Archive_ID())).archiveData(archiveBL.getBinaryData(arch)).build());
+	}
+
+	@Override
+	public void updatePrintedRecords(final ImmutableSet<ArchiveId> ids, final UserId userId)
+	{
+		queryBL.createQueryBuilder(I_AD_Archive.class, Env.getCtx(), ITrx.TRXNAME_None)
+				.addOnlyActiveRecordsFilter()
+				.addInArrayFilter(I_AD_Archive.COLUMN_AD_Archive_ID, ids).create().stream().filter(Objects::nonNull).forEach(archive -> archiveEventManager.firePdfUpdate(archive, userId, null));
```

Comment:
I guess you don't have access to `X_C_Doc_Outbound_Log_Line.ACTION_PdfExport` in here, right?
Still assuming `null => PdfExport` somewhere in the code is too dirty
So, suggestion: add a method `IArchiveEventManager.firePdfPdfExportUpdate(I_AD_Archive, UserId)`.
Then you can make the action `@NonNull` again

Reply:
In master `X_C_Doc_Outbound_Log_Line.ACTION_PdfExport` is presumed to be the default action. I guess this is legacy code, where the action is still passed in.
Will add an overload for firePdfUpdate.